### DISCUSSION TITLE
Add `timeout` promise function

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#203] [Ideas]
It's like our `delay` function, but it rejects upon completion.

## Usage

```ts
await Promise.race([
  someAsyncTask(),
  timeout(1000, "Optional message"),
])
```

The error looks like this:

```ts
Error { name: "TimeoutError", message: "Timed out: Optional message" }
```

**Comment by MarlonPassos-git** [comment]
I like the idea. It would be cool to be able to pass a class that extends the `Error` as a parameter, so that when the error is thrown, we could use our Custom Error. This helps a lot with error handling.

**Reply by aleclarson** [reply]
Good idea 😄 

Which signature should we do?

```ts
// A.
timeout(1000, "Optional message", MyError)

// B.
timeout(1000, MyError, "Optional message")

// C.
timeout(1000, () => new MyError("Optional message"))

// D.
timeout(1000, new MyError("Optional message"))
```

I would say D is out, since constructing errors pre-emptively is generally frowned upon for performance reasons, if I'm not mistaken.

The problem with B is that it requires more runtime type-checking if we also plan to support `timeout(1000, "optional message")` calls.

Of course, C is bit more verbose, but I'd say it reads nicer than A.

**Reply by MarlonPassos-git** [reply]
I prefer C. 

A and B can limit custom Erros with only one parameter.

**Comment by MarlonPassos-git** [comment]
https://github.com/radashi-org/radashi/pull/250
